File: python/book_manager.py
from typing import List
from book import Book

class BookManager:
    def __init__(self):
        self.books = []

    def add_book(self, book: Book) -> None:
        if book is None:
            raise ValueError("Buku tidak boleh kosong")
        self.books.append(book)

    def remove_book(self, title: str) -> bool:
        if title is None or not title.strip():
            raise ValueError("Judul tidak boleh kosong")

        title_lower = title.strip().lower()
        initial_count = len(self.books)
        self.books = [book for book in self.books if book.title.lower() != title_lower]
        return len(self.books) < initial_count

    def get_all_books(self) -> List[Book]:
        return self.books.copy()

    def find_books_by_author(self, author: str) -> List[Book]:
        if author is None or not author.strip():
            raise ValueError("Penulis tidak boleh kosong")

        author_lower = author.strip().lower()
        return [book for book in self.books if book.author.lower() == author_lower]

    def find_books_by_year(self, year: int) -> List[Book]:
        if year < 0 or year > 2100:
            raise ValueError("Tahun hanya bisa diisi dari tahun 2000 sampai 2100")
        return [book for book in self.books if book.year == year]

    def get_book_count(self) -> int:
        return len(self.books)

    def contains_book(self, title: str) -> bool:
        if title is None or not title.strip():
            raise ValueError("Judul tidak boleh kosong")

        title_lower = title.strip().lower()
        return any(book.title.lower() == title_lower for book in self.books)

    def clear_all_books(self) -> None:
        self.books.clear()

File: IdeaProjects/TugasPraktikum1_BookManager2/src/main/java/org/example/BookManager.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//

package org.example;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class BookManager {
    private final List<Book> buku = new ArrayList();

    public void addBook(Book book) {
        if (book == null) {
            throw new IllegalArgumentException("Buku tidak boleh kosong");
        } else {
            this.buku.add(book);
        }
    }

    public boolean removeBook(String judul) {
        if (judul != null && !judul.trim().isEmpty()) {
            return this.buku.removeIf((b) -> b.getJudul_Buku().equalsIgnoreCase(judul.trim()));
        } else {
            throw new IllegalArgumentException("Judul tidak boleh kosong");
        }
    }

    public List<Book> getAllBooks() {
        return new ArrayList(this.buku);
    }

    public List<Book> findBooksByAuthor(String penulis) {
        if (penulis != null && !penulis.trim().isEmpty()) {
            return (List)this.buku.stream().filter((book) -> book.getPenulis().equalsIgnoreCase(penulis.trim())).collect(Collectors.toList());
        } else {
            throw new IllegalArgumentException("Penulis tidak boleh kosong");
        }
    }

    public List<Book> findBooksByYear(int tahun) {
        if (tahun >= 2000 && tahun <= 2100) {
            return (List)this.buku.stream().filter((book) -> book.getTahun() == tahun).collect(Collectors.toList());
        } else {
            throw new IllegalArgumentException("Tahun hanya bisa diisi dari tahun 2000 sampai 2100");
        }
    }

    public int getBookCount() {
        return this.buku.size();
    }

    public boolean containsBook(String judul) {
        if (judul != null && !judul.trim().isEmpty()) {
            return this.buku.stream().anyMatch((book) -> book.getJudul_Buku().equalsIgnoreCase(judul.trim()));
        } else {
            throw new IllegalArgumentException("Judul tidak boleh kosong");
        }
    }

    public void clearAllBooks() {
        this.buku.clear();
    }
}
